diff -ruN parsec-3.0-orig/ext/splash2x/kernels/cholesky/src/malloc.C parsec-3.0-cholesky/ext/splash2x/kernels/cholesky/src/malloc.C
--- parsec-3.0-orig/ext/splash2x/kernels/cholesky/src/malloc.C	2012-09-04 22:56:31.000000000 -0400
+++ parsec-3.0-cholesky/ext/splash2x/kernels/cholesky/src/malloc.C	2013-03-04 18:27:17.539526680 -0500
@@ -20,6 +20,7 @@
 EXTERN_ENV
 
 #include "matrix.h"
+#include "tern/user.h"
 
 #define ALIGN 8
 #define MAXFAST 16
 
 
diff -ruN parsec-3.0-orig/ext/splash2x/kernels/cholesky/src/mf.C parsec-3.0-cholesky/ext/splash2x/kernels/cholesky/src/mf.C
--- parsec-3.0-orig/ext/splash2x/kernels/cholesky/src/mf.C	2012-09-04 22:56:31.000000000 -0400
+++ parsec-3.0-cholesky/ext/splash2x/kernels/cholesky/src/mf.C	2013-03-04 18:35:09.291510205 -0500
@@ -16,6 +16,7 @@
 
 EXTERN_ENV
 
+#include "tern/user.h"
 #include "matrix.h"
 #define HashNum 1024
 #define Bucket(desti, destj, src) ((desti+destj+src)%HashNum)
 
 
diff -ruN parsec-3.0-orig/ext/splash2x/kernels/cholesky/src/solve.C parsec-3.0-cholesky/ext/splash2x/kernels/cholesky/src/solve.C
--- parsec-3.0-orig/ext/splash2x/kernels/cholesky/src/solve.C	2012-09-04 22:56:31.000000000 -0400
+++ parsec-3.0-cholesky/ext/splash2x/kernels/cholesky/src/solve.C	2013-03-04 18:27:05.987527081 -0500
@@ -39,6 +39,7 @@
 
 #include <math.h>
 #include "matrix.h"
+#include "tern/user.h"
 
 #define SH_MEM_AMT   100000000
 #define DEFAULT_PPS         32
 
@@ -264,6 +267,7 @@
 	__parsec_roi_begin();
 #endif
 
+  soba_init(0, P, P*10000);
   CREATE(Go, P);
   WAIT_FOR_END(P);
 
