--- parsec-2.0-orig/pkgs/kernels/streamcluster/src/streamcluster.cpp	2009-01-18 19:34:18.000000000 -0500
+++ parsec-2.0/pkgs/kernels/streamcluster/src/streamcluster.cpp	2018-07-31 19:18:28.702968639 -0400
@@ -33,6 +33,8 @@
 #include <hooks.h>
 #endif
 
+#include "tern/user.h"
+
 using namespace std;
 
 #define MAXNAMESIZE 1024 // max filename length
@@ -1977,6 +1979,7 @@
     pthread_t* threads = new pthread_t[nproc];
     pkmedian_arg_t* arg = new pkmedian_arg_t[nproc];
 
+    slock_next_n(nproc);
     for( int i = 0; i < nproc; i++ ) {
       arg[i].points = points;
       arg[i].kmin = kmin;
